File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/model/storage/GeneralPreferencesTest.java
package com.dbottillo.mtgsearchfree.model.storage;

import android.annotation.SuppressLint;
import android.content.SharedPreferences;

import com.dbottillo.mtgsearchfree.util.AppInfo;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import static com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences.CARDS_SHOW_TYPE;
import static com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences.CARD_MIGRATION_REQUIRED;
import static com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences.DEBUG;
import static com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences.TOOLTIP_MAIN_SHOWN;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@SuppressLint("CommitPrefEdits")
public class GeneralPreferencesTest {

    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    private GeneralPreferences underTest;

    @Mock
    AppInfo appInfo;

    @Mock
    SharedPreferences sharedPreferences;

    @Mock
    SharedPreferences.Editor editor;

    @Before
    public void setup() {
        underTest = new GeneralPreferences(sharedPreferences, appInfo);
        when(sharedPreferences.edit()).thenReturn(editor);
        when(editor.putBoolean(any(String.class), any(Boolean.class))).thenReturn(editor);
        when(editor.putString(any(String.class), any(String.class))).thenReturn(editor);
    }

    @Test
    public void setDebug_shouldUpdateSharedPreferences() {
        underTest.setDebug();

        verify(sharedPreferences).edit();
        verify(editor).putBoolean(DEBUG, true);
        verify(editor).apply();
    }

    @Test
    public void setCardsShowTypeList_shouldUpdateSharedPreferences() {
        underTest.setCardsShowTypeList();

        verify(sharedPreferences).edit();
        verify(editor).putString(CARDS_SHOW_TYPE, "List");
        verify(editor).apply();
    }

    @Test
    public void setCardsShowTypeGrid_shouldUpdateSharedPreferences() {
        underTest.setCardsShowTypeGrid();

        verify(sharedPreferences).edit();
        verify(editor).putString(CARDS_SHOW_TYPE, "Grid");
        verify(editor).apply();
    }

    @Test
    public void setTooltipMainHide_shouldUpdateSharedPreferences() {
        underTest.setTooltipMainHide();

        verify(sharedPreferences).edit();
        verify(editor).putBoolean(TOOLTIP_MAIN_SHOWN, false);
        verify(editor).apply();
    }

    @Test
    public void isTooltipMainToShow_shouldReturnTrueIfItsNotAFreshInstall_AndSharedPreferencesIsTrue() {
        when(sharedPreferences.getBoolean(TOOLTIP_MAIN_SHOWN, true)).thenReturn(true);
        when(appInfo.getFirstInstallTime()).thenReturn(200L);
        when(appInfo.getLastUpdateTime()).thenReturn(400L);

        assertTrue(underTest.isTooltipMainToShow());
    }

    @Test
    public void isTooltipMainToShow_shouldReturnFalseIfItsAFreshInstall() {
        when(appInfo.getFirstInstallTime()).thenReturn(200L);
        when(appInfo.getLastUpdateTime()).thenReturn(200L);

        assertFalse(underTest.isTooltipMainToShow());
    }

    @Test
    public void isTooltipMainToShow_shouldReturnFalseIfItsNotAFreshInstall_AnsSharedPreferencesIsFalse() {
        when(sharedPreferences.getBoolean(TOOLTIP_MAIN_SHOWN, true)).thenReturn(false);
        when(appInfo.getFirstInstallTime()).thenReturn(200L);
        when(appInfo.getLastUpdateTime()).thenReturn(400L);

        assertFalse(underTest.isTooltipMainToShow());
    }

    @Test
    public void markCardMigrationStarted_shouldUpdateSharedPreferences() {
        underTest.markCardMigrationStarted();

        verify(sharedPreferences).edit();
        verify(editor).putBoolean(CARD_MIGRATION_REQUIRED, false);
        verify(editor).apply();
    }

}

File: MTGSearch/src/test/kotlin/com/dbottillo/mtgsearchfree/model/storage/GeneralPreferencesTest.kt
package com.dbottillo.mtgsearchfree.model.storage

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.Editor
import com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences.Companion.CARDS_SHOW_TYPE
import com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences.Companion.CARD_MIGRATION_REQUIRED
import com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences.Companion.DEBUG
import com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences.Companion.TOOLTIP_MAIN_SHOWN
import com.dbottillo.mtgsearchfree.util.AppInfo
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnit

@SuppressLint("CommitPrefEdits")
class GeneralPreferencesTest {

    @Rule
    @JvmField
    var mockitoRule = MockitoJUnit.rule()

    lateinit var underTest: GeneralPreferences

    @Mock
    lateinit var appInfo: AppInfo

    @Mock
    lateinit var sharedPreferences: SharedPreferences

    @Mock
    lateinit var appContext: Context

    @Mock
    lateinit var editor: Editor

    @Before
    fun setup() {
        whenever(appContext.getSharedPreferences("General", Context.MODE_PRIVATE)).thenReturn(sharedPreferences)
        underTest = GeneralPreferences(appContext, appInfo)
        `when`(sharedPreferences.edit()).thenReturn(editor)
        `when`(editor.putBoolean(anyString(), anyBoolean())).thenReturn(editor)
        `when`(editor.putString(anyString(), anyString())).thenReturn(editor)
    }

    @Test
    fun setDebug_shouldUpdateSharedPreferences() {
        underTest.setDebug()

        verify(sharedPreferences).edit()
        verify(editor).putBoolean(DEBUG, true)
        verify(editor).apply()
    }

    @Test
    fun setCardsShowTypeList_shouldUpdateSharedPreferences() {
        underTest.setCardsShowTypeList()

        verify(sharedPreferences).edit()
        verify(editor).putString(CARDS_SHOW_TYPE, "List")
        verify(editor).apply()
    }

    @Test
    fun setCardsShowTypeGrid_shouldUpdateSharedPreferences() {
        underTest.setCardsShowTypeGrid()

        verify(sharedPreferences).edit()
        verify(editor).putString(CARDS_SHOW_TYPE, "Grid")
        verify(editor).apply()
    }

    @Test
    fun setTooltipMainHide_shouldUpdateSharedPreferences() {
        underTest.setTooltipMainHide()

        verify(sharedPreferences).edit()
        verify(editor).putBoolean(TOOLTIP_MAIN_SHOWN, false)
        verify(editor).apply()
    }

    @Test
    fun isTooltipMainToShow_shouldReturnTrueIfItsNotAFreshInstall_AndSharedPreferencesIsTrue() {
        `when`(sharedPreferences.getBoolean(TOOLTIP_MAIN_SHOWN, true)).thenReturn(true)
        `when`(appInfo.firstInstallTime).thenReturn(200L)
        `when`(appInfo.lastUpdateTime).thenReturn(400L)

        assertTrue(underTest.isTooltipMainToShow())
    }

    @Test
    fun isTooltipMainToShow_shouldReturnFalseIfItsAFreshInstall() {
        `when`(appInfo.firstInstallTime).thenReturn(200L)
        `when`(appInfo.lastUpdateTime).thenReturn(200L)

        assertFalse(underTest.isTooltipMainToShow())
    }

    @Test
    fun isTooltipMainToShow_shouldReturnFalseIfItsNotAFreshInstall_AnsSharedPreferencesIsFalse() {
        `when`(sharedPreferences.getBoolean(TOOLTIP_MAIN_SHOWN, true)).thenReturn(false)
        `when`(appInfo.firstInstallTime).thenReturn(200L)
        `when`(appInfo.lastUpdateTime).thenReturn(400L)

        assertFalse(underTest.isTooltipMainToShow())
    }

    @Test
    fun markCardMigrationStarted_shouldUpdateSharedPreferences() {
        underTest.markCardMigrationStarted()

        verify(sharedPreferences).edit()
        verify(editor).putBoolean(CARD_MIGRATION_REQUIRED, false)
        verify(editor).apply()
    }

}